File: spiralOrderGenerateII.java
//M*N matrix
//soluiton: iterative
public int[][] spiralGenerateII(int m, int n) {
	int[][] matrix = new int[m][n];
	if(n == 0 || m == 0) 
		return matrix;
	int num = 1;
	int left = 0;
	int right = n-1;
	int up = 0;
	int down = m-1;
	while(left < right && up < down){
		for(int i = left; i <= right; i++)
			matrix[up][i] = num++;
		for(int i = up+1; i <= down-1; i++)
			matrix[i][right] = num++;
		for(int i = right; i >= left; i--)
			matrix[down][i] = num++;
		for(int i = down-1; i >= up +1; i--)
			matrix[i][left] = num++;
		left++;
		right--;
		up++;
		down--;
	}
	if(left > right || up > down)
		return matrix;
	if(left == right){
		for(int i = up; i <= down; i++)
			matrix[i][left] = num++;
	}else{
		for(int i = left; i <= right; i++)
			matrix[up][i] = num++;
	}
	return matrix;
}

File: matrix/spiralOrderGenerateII.py
def spiralGenerate1(m,n):
	matrix = [[0]*n for i in range(m)]   #n col and m row
	num = 1
	left = 0
	right = n-1
	up = 0
	right = m-1

	while left < right and up < down:
		for i in range(left, right):
			matrix[left][i] = num
			num += 1
			i += 1
		for i i range(up, down):
			matrix[i][right] = num
			num += 1
			i += 1
		for i in range(right, left, -1):
			matrix[down][i] = num
			num += 1
			i -= 1
		for i in range(down, up, -1):
			matrix[i][left] = num
			num += 1
			i -= 1

		left += 1
		right -= 1
		up += 1
		down -= 1

	if left > right or up > down:
		return matrix

	if left == right:
		for i in range(up, down+1):
			matrix[up][i] = num
			num += 1
			i+= 1

	else:
		for i in range(left, right+1):
			matrix[i][left] = num
			num += 1
			i += 1

	return matrix
